### PR TITLE
Remove inclusive ranges

### LINKED ISSUE #586: proc maps ranges are not inclusive
Today I discovered the range-map code in this repo.
It uses a `Range` type which models an **inclusive** range.

I don't think the code in this repo is very well served by an inclusive range type. I see lots and lots of occurrences of the pattern "end_address - 1".

Here's a place where we *don't* subtract 1 from the end, and I think that's wrong:

https://github.com/rust-minidump/rust-minidump/blob/fd680d28262bc84bdca6c9e4f64f1d3d937f081b/minidump/src/minidump.rs#L2337-L2343

Searching the web for "proc maps" turns up examples such as this one:

```
7faa721ce000-7faa721cf000 r--p 00000000 08:30 3021                       /usr/lib/locale/C.UTF-8/LC_ADDRESS
7faa721cf000-7faa724b5000 r--p 00000000 08:30 3034                       /usr/lib/locale/locale-archive
7faa724b5000-7faa724da000 r--p 00000000 08:30 11971                      /usr/lib/x86_64-linux-gnu/libc-2.31.so
```

These ranges are consecutive; the end address of one range is equal to the start address of the next range. The ranges don't overlap.

So really I'm suggesting two things in this bug:

 - We should fix the proc maps code, and
 - Inclusive ranges of integers are basically never what you want.

How do people feel about switching to the `rangemap` crate (no dash between "range" and "map")? https://docs.rs/rangemap/

### PR DESCRIPTION
**Pull request by mstange**
Fixes #586.

This PR removes the dependency on range_map, which allows us to stop using inclusive ranges.

All of our ranges describe ranges of bytes in memory; for those, you basically always want half-open ranges of the form `start_address..end_address`.

To replace `range_map::RangeMap`, we use either sorted Vecs, or `rangemap::RangeMap`.
Note the missing underscore.

`rangemap::RangeMap` does not implement PartialEq / Eq, so some of the tests are changed from `assert_eq!` to a match statement.

It also does not support double-ended iterators.

For any RangeMaps which I converted to Vecs, I've changed the way we handle overlapping ranges, to do the following:

```
cur.lines.sort_by_key(|l| l.address);
cur.lines.dedup_by(|next, current| current.end_address() >= next.end_address());
```

This allows overlapping ranges if the next range sticks out from the end of its predecessor. And I've changed the tests to allow either value to be returned in the overlapping part.
Since overlapping ranges are usually a garbage input anyway, I'm hoping that this is an adequate way to deal with the situation. From what I gathered from the history of this code, the way we dealt with overlapping ranges was mostly aimed at avoiding panics from range_map, and not with the goal of any particular output.

There are two case where we care about the specific way we process overlapping ranges is this case, which I've preserved. The first one is about STACK WIN records:

https://github.com/rust-minidump/rust-minidump/blob/e0f8b5a7b4b6d9eebdf297f59c7aaa49c93f8494/breakpad-symbols/src/sym_file/parser.rs#L443-L474

And the second is about duplicated modules, and is handled in this breakpad code: https://chromium.googlesource.com/breakpad/breakpad/+/54fa71efbe50fb2b58096d871575b59e12edba6d/src/processor/minidump.cc#2784
I'm pretty sure I preserved that one too.

Overall, it's not the prettiest change, especially because I'm adding multiple manual calls to binary_search. But at least it's getting rid of a lot of occurrences of the pattern `end_address - 1`.

### CODE REVIEW

File: breakpad-symbols/src/lib.rs [hunk 1]
```diff
@@ -895,10 +895,10 @@ mod test {
...
         let supplier = SimpleSymbolSupplier::new(paths.clone());
         let bad = SimpleModule::default();
-        assert_eq!(
-            supplier.locate_symbols(&bad).await,
-            Err(SymbolError::NotFound)
-        );
+        match supplier.locate_symbols(&bad).await {
+            Err(SymbolError::NotFound) => {}
+            _ => panic!(),
+        }
```
Comment (Gankra) [inline breakpad-symbols/src/lib.rs:901]: these might be clearer as `assert!(matches!(expr, Err(SymbolError::NotFound))`

File: breakpad-symbols/src/sym_file/parser.rs [hunk 2]
```diff
@@ -3,19 +3,31 @@
...
+trait RangeIntersects {
+    fn intersects(&self, other: &Self) -> bool;
+}
+
+impl<T: Ord> RangeIntersects for Range<T> {
+    fn intersects(&self, other: &Self) -> bool {
+        self.start < self.end
+            && other.start < other.end
+            && self.start <= other.end
+            && other.start < self.end
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:26]: I think this merits a bit of a comment. The first two checks just seem to be "this isn't an empty (or inverted) range" but I'm a bit confused by why one of the comparisons is <=...

File: breakpad-symbols/src/sym_file/parser.rs [hunk 3]
```diff
@@ -510,21 +521,18 @@ impl SymbolParser {
...
                         // Line data from PDB files often has a zero-size line entry, so just
                         // filter those out.
-                        if l.size > 0 {
-                            if let Some(end) = l.address.checked_add(l.size as u64 - 1) {
-                                return (Some(Range::new(l.address, end)), l);
-                            }
-                        }
-
-                        (None, l)
+                        l.size > 0
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:527]: I was like "why the fuck did I write this like this" and tried to follow the blame and I think this is just ancient code that survived a million refactors.

File: breakpad-symbols/src/sym_file/parser.rs [hunk 4]
```diff
@@ -1111,39 +1078,40 @@ FUNC 1001 10 10 some func overlap contained
...
-        .ranges_values()
-        .map(|&(_, ref f)| f)
-        .collect::<Vec<_>>();
+    assert_eq!(sym.functions.len(), 2);
+    assert!(sym
+        .function_for_address(0x1000)
+        .unwrap()
+        .name
+        .starts_with("some func"));
+    assert_eq!(sym.function_for_address(0x1030).unwrap().address, 0x1001);
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:1087]: ! what's up with this

File: breakpad-symbols/src/sym_file/parser.rs [hunk 5]
```diff
@@ -1111,39 +1078,40 @@ FUNC 1001 10 10 some func overlap contained
...
+                    size: 0x10,
+                    file: 0,
+                    line: 42,
+                },
+                SourceLine {
+                    address: 0x1001,
+                    size: 0x10,
+                    file: 0,
+                    line: 44,
+                },
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:1108]: ditto

File: breakpad-symbols/src/sym_file/parser.rs [hunk 6]
```diff
@@ -1286,7 +1254,7 @@ STACK WIN 4 8d93e 4 4 0 0 10 0 0 1 prog string
 fn address_size_overflow() {
     let bytes = b"FUNC 1 2 3 x\nffffffffffffffff 2 0 0\n";
     let sym = parse_symbol_bytes(bytes.as_slice()).unwrap();
-    let fun = sym.functions.get(1).unwrap();
-    assert!(fun.lines.is_empty());
+    let fun = sym.function_for_address(1).unwrap();
+    assert_eq!(fun.line_for_address(0x1234), None);
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/parser.rs:1258]: ? why the change?

File: breakpad-symbols/src/sym_file/types.rs [hunk 7]
```diff
@@ -33,6 +34,12 @@ pub struct SourceLine {
     pub line: u32,
 }
 
+impl SourceLine {
+    pub fn end_address(&self) -> u64 {
+        self.address.saturating_add(self.size.into())
+    }
+}
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/types.rs:41]: hmm saturating always feels like a trap, why do you think this is better than discarding the entry?

File: breakpad-symbols/src/sym_file/types.rs [hunk 8]
```diff
@@ -45,18 +52,34 @@ pub struct Function {
...
         }
-        Some(Range::new(
-            self.address,
-            self.address.checked_add(self.size as u64)? - 1,
-        ))
+        Some(self.address..self.end_address())
+    }
+
+    pub fn line_for_address(&self, address: u64) -> Option<&SourceLine> {
+        match self.lines.binary_search_by_key(&address, |l| l.address) {
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/types.rs:71]: I would prefer if you could transplant the comment you deleted, binary-search is hell and always merits comments.

(edit: oh wait you added some to the other instance, this probably still merits them though)

File: breakpad-symbols/src/sym_file/types.rs [hunk 9]
```diff
@@ -96,10 +119,7 @@ impl StackInfoCfi {
         if self.size == 0 {
             return None;
         }
-        Some(Range::new(
-            self.init.address,
-            self.init.address.checked_add(self.size as u64)? - 1,
-        ))
+        Some(self.init.address..self.init.address.checked_add(self.size as u64)?)
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/types.rs:122]: this one stays checked while all the others saturate?

File: breakpad-symbols/src/sym_file/types.rs [hunk 10]
```diff
@@ -149,22 +169,19 @@ impl StackInfoWin {
         if self.size == 0 {
             return None;
         }
-        Some(Range::new(
-            self.address,
-            self.address.checked_add(self.size as u64)? - 1,
-        ))
+        Some(self.address..self.address.checked_add(self.size as u64)?)
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/types.rs:172]: Ah another case where you check still

File: breakpad-symbols/src/sym_file/types.rs [hunk 11]
```diff
@@ -222,3 +239,27 @@ pub struct SymbolFile {
     /// still be correct.
     pub cfi_eval_corruptions: u64,
 }
+
+impl SymbolFile {
+    /// Returns the closest function to the address for which `function.address <= address`.
+    pub fn find_nearest_function(&self, address: u64) -> Option<&Function> {
```
Comment (Gankra) [inline breakpad-symbols/src/sym_file/types.rs:245]: Feels like this wants a better name. `find_nearest_function_before`?

**Review by Gankra — CHANGES_REQUESTED** [review]
Reviewed everything except minidump.rs, taking a break, this is a huge diff @-@

File: breakpad-symbols/src/sym_file/parser.rs (hunk 4 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/parser.rs:1087]: Yeah this is probably the most major change in behavior. We now allow overlapping functions if they aren't fully subsumed. And for the addresses in the overlapping parts, we don't care which function is returned. In the first test, all of the candidates begin with "some func". And the second one is testing that the function which starts at 0x1001 and sticks out the end from the previous one actually covers the byte at 0x1030.

File: breakpad-symbols/src/lib.rs (hunk 1 above)
Comment (mstange) [reply breakpad-symbols/src/lib.rs:901]: Oh I didn't realize `matches!` was a thing! All of its friends (e.g. assert_matches) seemed to be unstable so I didn't look further.

File: breakpad-symbols/src/sym_file/parser.rs (hunk 2 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/parser.rs:26]: You caught a bug, the <= is wrong. I was following [our rectangle intersection code](https://searchfox.org/mozilla-central/rev/0069a0cdc3cd136ea7411521219df6665f260be7/gfx/2d/BaseRect.h#105-107) and still screwed it up.

File: breakpad-symbols/src/sym_file/types.rs (hunk 7 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/types.rs:41]: I don't have strong opinions on this. I used saturating simply because then I don't have to deal with None.
We may produce garbage in the bad cases but at least we won't panic.

File: breakpad-symbols/src/sym_file/types.rs (hunk 8 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/types.rs:71]: ok, will make sure there are some comments at every instance of this pattern

File: breakpad-symbols/src/sym_file/types.rs (hunk 9 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/types.rs:122]: Hmm, true, it's inconsistent that I kept all the methods which return ranges to use checked, and then used saturate in `end_address()`. The methods which return a range want to return None if the range is empty or inverted, so using checked is easier there.

File: breakpad-symbols/src/sym_file/types.rs (hunk 11 above)
Comment (mstange) [reply breakpad-symbols/src/sym_file/types.rs:245]: Sure, sounds good.

File: breakpad-symbols/src/sym_file/types.rs (hunk 7 above)
Comment (Gankra) [reply breakpad-symbols/src/sym_file/types.rs:41]: My default calculus here is that one Bad input can spoil the batch if we keep it around, but discarding a sketchy input (as in discarding the line) is unlikely to even be noticed, since we have millions of lines and you're gonna query like 20 of them. The places where I go out of my way to preserve sketchy inputs are places where it was an actual problem in production.

Comment (mstange) [reply breakpad-symbols/src/sym_file/types.rs:41]: Filtering them out sounds good to me. But we use the same type for unfiltered and filtered line records. Should we return Option from end_address or not?

File: minidump/src/minidump.rs [hunk 12]
```diff
@@ -1397,19 +1384,23 @@ impl MinidumpModuleList {
...
-            .enumerate()
-            .map(|(i, module)| (module.memory_range(), i))
-            .into_rangemap_safe();
+        let mut sorted_modules: Vec<usize> = (0..modules.len()).collect();
+        sorted_modules.sort_by_key(|i| modules[*i].base_address());
+        sorted_modules.dedup_by(|ni, ci| {
+            let cur = &modules[*ci];
+            let next = &modules[*ni];
+            let cur_end_address = cur.base_address().saturating_add(cur.size());
+            let next_end_address = next.base_address().saturating_add(next.size());
```
Comment (Gankra) [inline minidump/src/minidump.rs:1398]: More saturating that should probably be checked. This one's harder to remove though, so maybe it's fine?

File: minidump/src/minidump.rs [hunk 13]
```diff
@@ -1426,9 +1417,22 @@ impl MinidumpModuleList {
...
+            .sorted_modules
+            .binary_search_by_key(&address, |i| self.modules[*i].base_address())
+        {
+            Ok(i) => &self.modules[self.sorted_modules[i]],
+            Err(0) => return None,
+            Err(i) => &self.modules[self.sorted_modules[i - 1]],
+        };
+        let end_address = nearest_module
+            .base_address()
+            .saturating_add(nearest_module.size());
```
Comment (Gankra) [inline minidump/src/minidump.rs:1430]: in theory this one should be validated on first digest so it's fine for it to panic..?

File: minidump/src/minidump.rs [hunk 14]
```diff
@@ -1426,9 +1417,22 @@ impl MinidumpModuleList {
 
     /// Return a `MinidumpModule` whose address range covers `address`.
     pub fn module_at_address(&self, address: u64) -> Option<&MinidumpModule> {
-        self.modules_by_addr
-            .get(address)
-            .map(|&index| &self.modules[index])
+        let nearest_module = match self
+            .sorted_modules
+            .binary_search_by_key(&address, |i| self.modules[*i].base_address())
```
Comment (Gankra) [inline minidump/src/minidump.rs:1422]: wait do we really get weird overlapping modules? gross that this can't use rangemap :(

File: minidump/src/minidump.rs [hunk 15]
```diff
@@ -1721,22 +1720,26 @@ impl<'mdmp, Descriptor> MinidumpMemoryListBase<'mdmp, Descriptor> {
...
-            .enumerate()
-            .map(|(i, region)| (region.memory_range(), i))
-            .into_rangemap_safe();
+        let mut sorted_regions: Vec<usize> = (0..regions.len()).collect();
+        sorted_regions.sort_by_key(|i| regions[*i].base_address);
+        sorted_regions.dedup_by(|ni, ci| {
+            let cur = &regions[*ci];
+            let next = &regions[*ni];
+            let cur_end_address = cur.base_address.saturating_add(cur.size);
+            let next_end_address = next.base_address.saturating_add(next.size);
```
Comment (Gankra) [inline minidump/src/minidump.rs:1737]: more saturating that's annoying to remove

File: minidump/src/minidump.rs [hunk 16]
```diff
@@ -1745,9 +1748,22 @@ impl<'mdmp, Descriptor> MinidumpMemoryListBase<'mdmp, Descriptor> {
...
+            Err(0) => return None,
+            Err(i) => &self.regions[self.sorted_regions[i - 1]],
+        };
+        let end_address = nearest_region
+            .base_address
+            .saturating_add(nearest_region.size);
+        if address < end_address {
+            Some(nearest_region)
+        } else {
+            None
```
Comment (Gankra) [inline minidump/src/minidump.rs:1765]: geez this pattern gets repeated a lot :(

**Review by Gankra — CHANGES_REQUESTED** [review]
Ok review of minidump.rs... not a huge fan of how much manual binary searching is going on here.

File: minidump/src/minidump.rs (hunk 16 above)
Comment (Gankra) [reply minidump/src/minidump.rs:1765]: (the binary searching)

**Comment by Gankra** [comment]
we should add a counter for "times people have tried to remove range-map and given up" D:
